Factor various logging methods in DataFlash

peterbarker opened the pull request:
Also fixes two bugs:
 - incorrect packet type used for battery cells
 - IMUDT dleta-velocity incorrect for third IMU

OXINARF commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1545,71 +1545,54 @@ void DataFlash_Class::Log_Write_AttitudeView(AP_AHRS_View &ahrs, const Vector3f
+    bool has_temp = battery.get_temperature(temp, battery_instance);
+    struct log_Current pkt = {
+        LOG_PACKET_HEADER_INIT(type),
+        time_us             : AP_HAL::micros64(),
```
Shouldn't this be `time_us`?

OXINARF commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1545,71 +1545,54 @@ void DataFlash_Class::Log_Write_AttitudeView(AP_AHRS_View &ahrs, const Vector3f
+        voltage_resting     : battery.voltage_resting_estimate(battery_instance),
+        current_amps        : battery.current_amps(battery_instance),
+        current_total       : battery.current_total_mah(battery_instance),
+        temperature         : (int16_t)(has_temp ? (temp * 100) : battery_instance),
```
This `battery_instance` is wrong, it should stay a 0. That search and replace failed this time 😜

OXINARF commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -463,17 +463,17 @@ void DataFlash_Class::Log_Write_IMU(const AP_InertialSensor &ins)
 
 // Write an accel/gyro delta time data packet
-void DataFlash_Class::Log_Write_IMUDT(const AP_InertialSensor &ins, uint64_t time_us, uint8_t imu_mask)
+void DataFlash_Class::Log_Write_IMUDT_instance(const AP_InertialSensor &ins, const uint64_t time_us, const uint8_t imu_instance, const enum LogMessages type)
```
I like the added const, can we do that for the other methods as well?

OXINARF commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1575,85 +1509,73 @@ void DataFlash_Class::Log_Write_AttitudeView(AP_AHRS_View &ahrs, const Vector3f
+        const AP_BattMonitor::cells &cells = battery.get_cell_voltages(battery_instance);
+        struct log_Current_Cells cell_pkt = {
+            LOG_PACKET_HEADER_INIT(celltype),
             time_us             : AP_HAL::micros64(),
```
Shouldn't this be `time_us` too?

peterbarker commented:
On Fri, 29 Sep 2017, Francisco Ferreira wrote:

> Shouldn't this be time_us?

Yep, fixed.

> +        temperature         : (int16_t)(has_temp ? (temp * 100) : battery_i
> nstance),
> 
> This battery_instance is wrong, it should stay a 0. That search and replace
> failed this time 😜

Fixed.  You *do* see the 00 in "100" I *didn't* screw up, right? :-)  I 
claim they distracted me.

> In libraries/DataFlash/LogFile.cpp:
> 
> > @@ -463,17 +463,17 @@ void DataFlash_Class::Log_Write_IMU(const AP_Inertia
> lSensor &ins)
>  }
>
>  // Write an accel/gyro delta time data packet
> -void DataFlash_Class::Log_Write_IMUDT(const AP_InertialSensor &ins, uint64_
> t time_us, uint8_t imu_mask)
> +void DataFlash_Class::Log_Write_IMUDT_instance(const AP_InertialSensor &ins
> , const uint64_t time_us, const uint8_t imu_instance, const enum LogMessages
>  type)
> 
> I like the added const, can we do that for the other methods as well?

I think I did manage to get all the ones I modified in this PR?

> Shouldn't this be time_us too?

Yep, fixed.  That was a small functionality change I slipped in 
considering comments on the baro PR.

OXINARF commented:
> I think I did manage to get all the ones I modified in this PR?

Sorry, I should have been explicit. I was referring to `Log_Write_IMU_instance` and `Log_Write_Compass_instance`.
By the way, a rebase on master is appreciated 🙂

OXINARF commented:
Ah, it's just the declaration in the header missing it!

peterbarker commented:
On Fri, 29 Sep 2017, Francisco Ferreira wrote:

>       I think I did manage to get all the ones I modified in this PR?
> 
> Sorry, I should have been explicit. I was referring to
> Log_Write_IMU_instance and Log_Write_Compass_instance.
> By the way, a rebase on master is appreciated 🙂

Ah, those.

So the definitions include the consts on all parameters.  The declarations 
don't - except where I obviously screwed up, in current and imudt.

I'm really not fussed whether they go into the definition or not, but 
I've heard people complain they really only add noise for the simple types 
which are just going on the stack anyway.  I do habitually add them into 
the definitions because they do help catch real bugs.

You're suggesting you'd like to see consts in definitions where possible? 
Not too noisy?

peterbarker commented:
On Fri, 29 Sep 2017, Francisco Ferreira wrote:

> Ah, it's just the declaration in the header missing it!

Rebase done.

OXINARF commented:
> You're suggesting you'd like to see consts in definitions where possible?
Not too noisy?

I like to always have declaration and definition match. But most of all I like coherence, so either all or none with const in declaration, I'll take it.

OXINARF commented:
@peterbarker No need to argue about it, I understand your points, but I still like to have declaration and definition match. As I said before, I don't really care what goes in as long as all methods do the same, instead of the current code:

<pre>
void Log_Write_IMU_instance(const AP_InertialSensor &ins,
                            <b>uint64_t</b> time_us,
                            <b>uint8_t</b> imu_instance,
                            <b>enum LogMessages</b> type);

void Log_Write_Compass_instance(const Compass &compass,
                                <b>uint64_t</b> time_us,
                                <b>uint8_t</b> mag_instance,
                                <b>enum LogMessages</b> type);

void Log_Write_Current_instance(const AP_BattMonitor &battery,
                                <b>const uint64_t</b> time_us,
                                <b>const uint8_t</b> battery_instance,
                                <b>const enum LogMessages</b> type,
                                <b>const enum LogMessages</b> celltype);

void Log_Write_IMUDT_instance(const AP_InertialSensor &ins,
                              <b>const uint64_t</b> time_us,
                              <b>const uint8_t</b> imu_instance,
                              <b>const enum LogMessages</b> type);
</pre>

I see absolutely no reason why some methods have const and others don't.

peterbarker commented:
@OXINARF OK, finally got it.  The short words helped :-)

I've pushed up declarations which are consistent.

OXINARF commented:
Merged, thanks!
